### PR TITLE
Re-introduce updated architecture overview diagram

### PR DESCRIPTION
**Pull request by tsenart**
This commit re-introduces an initial updated architecture overview diagram produced with Graphviz,  with the intent of aiding the on-boarding of new team mates.

architecture

We should try to keep this up to date, so whenever a new service or link between services happens, try to remember to update this diagram. It's hard to enforce this so I think we need to rely on building this habit, so I updated the GitHub PR template to that effect.

@sourcegraph/code-intel @sourcegraph/core-services @sourcegraph/web @sourcegraph/distribution: Please suggest changes, updates, refinements. Thanks!

### CODE REVIEW

**Comment by efritz** [comment]
`lsif-{server,worker}` read/write to both `postgres` and `redis-store`. I can update the diagram if you'd like.

**Comment by uwedeportivo** [comment]
is query-runner the only node that uses the frontend-internal API. i thought that more nodes ask frontend-internal questions instead of going to the DB directly (i just had the discussion with stephen)

**Comment by tsenart** [comment]
> lsif-{server,worker} read/write to both postgres and redis-store. I can update the diagram if you'd like.

Please do! About to hop off for the weekend :)

**Comment by tsenart** [comment]
> is query-runner the only node that uses the frontend-internal API. i thought that more nodes ask frontend-internal questions instead of going to the DB directly (i just had the discussion with stephen)

It'd be great to find out and update The Graph ❤️

**Comment by efritz** [comment]
@tsenart What are single vs double arrows for?

**Comment by slimsag** [comment]
Some missing things I am certain about:

1. The management console is missing, it talks *just* to the postgres DB.
2. Replacer, searcher, repo-updater, lsif-server and lsif-worker all have bi-directional communication with the frontend.
3. "Clients" should include "Customer scripts" or something.
4. the Symbols service is missing, it has bi-directional communication with the frontend. It also talks to gitserver.

**Comment by slimsag** [comment]
@efritz can you update for #5 in my list above if needed

**Comment by efritz** [comment]
@slimsag I did group them anyway because they are tightly coupled and it makes sense to think of them as one unit.

**Comment by nicksnyder** [comment]
Thanks for doing this! zoekt-indexserver is in a weird place. Is it possible to group it with zoekt-webserver (e.g. "Indexed search") similar to how code hosts are grouped?

**Comment by efritz** [comment]
@nicksnyder @slimsag updated to address all your comments

**Comment by uwedeportivo** [comment]
would it be possible to distinguish nodes on whether they scale horizontally or not: so have zoekt and gitserver marked with shadow shapes or something and have the singletons be like they are now ?

**Comment by slimsag** [comment]
If answer to @uwedeportivo's question is yes, the horizontally scalable ones are:

- LSIF
- searcher
- replacer
- symbols
- indexed-search
- gitserver
-

File: doc/dev/architecture/architecture.dot [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    gitserver -> gitlab_dot_com
+    gitserver -> bitbucket_server
+    gitserver -> bitbucket_cloud
+
+    zoekt_indexserver -> frontend
+    zoekt_indexserver -> gitserver
+
+    browser_ext -> frontend
+    web_app -> frontend
+    src_cli -> frontend
```
Comment (felixfbecker) [inline doc/dev/architecture/architecture.dot]: Isn't customer scripts missing a connection to frontend here?

File: doc/dev/architecture/architecture.dot [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    repo_updater [label="repo-updater" URL="https://github.com/sourcegraph/sourcegraph/tree/master/cmd/repo-updater"]
+    searcher     [label="searcher" URL="https://github.com/sourcegraph/sourcegraph/tree/master/cmd/searcher"]
+    replacer     [label="replacer" URL="https://github.com/sourcegraph/sourcegraph/tree/master/cmd/replacer"]
+    query_runner [label="query-runner" URL="https://github.com/sourcegraph/sourcegraph/tree/master/cmd/query-runner"]
+    github_proxy [label="github-proxy" URL="https://github.com/sourcegraph/sourcegraph/tree/master/cmd/github-proxy"]
+    syntect_server [label="syntect-server" URL="https://github.com/sourcegraph/syntect_server"]
+    management_console [label="management-console" URL="https://github.com/sourcegraph/sourcegraph/tree/master/cmd/management-console"]
+    symbols [label="symbols" URL="https://github.com/sourcegraph/sourcegraph/tree/master/cmd/symbols"]
+
+    subgraph cluster_lsif_zoekt {
```
Comment (tsenart) [inline doc/dev/architecture/architecture.dot]:
```suggestion
    subgraph cluster_zoekt {
```

File: doc/dev/architecture/architecture.dot [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        label="Indexed search"
+        rank="same"
+        graph [style="dotted"]
+
+        zoekt_webserver [label="zoekt-webserver" URL="https://github.com/sourcegraph/zoekt/tree/master/cmd/zoekt-webserver"]
+        zoekt_indexserver [label="zoekt-indexserver" URL="https://github.com/sourcegraph/zoekt/tree/master/cmd/zoekt-sourcegraph-indexserver"]
+    }
+
+    subgraph cluster_lsif_processes {
```
Comment (tsenart) [inline doc/dev/architecture/architecture.dot]:
```suggestion
    subgraph cluster_lsif {
```

**Comment by tsenart** [comment]
> @tsenart What are single vs double arrows for?

Directionality. Single arrow means origin node issues requests to the destination node. Double arrow means both do.

**Comment by tsenart** [comment]
> would it be possible to distinguish nodes on whether they scale horizontally or not: so have zoekt and gitserver marked with shadow shapes or something and have the singletons be like they are now ?

I think that's a great dimension to capture here. I'll sort something out!

**Comment by tsenart** [comment]
> Replacer, searcher, repo-updater, lsif-server and lsif-worker all have bi-directional communication with the frontend.

@slimsag: I guess repo-updater issues requests to the frontend only because of the omnipresent `conf` package. Data layer is done directly with postgres these days.

File: doc/dev/architecture/architecture.dot (hunk 1 above)
Comment (tsenart) [reply doc/dev/architecture/architecture.dot]: Feel free to add it if it's missing! I'll take a look at it on Monday, and will add if you didn't in the meantime.

**Comment by slimsag** [comment]
> @slimsag: I guess repo-updater issues requests to the frontend only because of the omnipresent conf package. Data layer is done directly with postgres these days.

Yes, correct

**Comment by tsenart** [comment]
What do you think of this top down version (rather than left to right)?
horizontal

**Comment by uwedeportivo** [comment]
i like it

File: doc/dev/architecture/architecture.dot [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    subgraph cluster_codehosts {
+        label="Code hosts"
+        rank="same"
+        graph [style="dotted"]
+        node  [shape="oval"]
+
+        github_dot_com [label="github.com"]
+        gitlab_dot_com [label="gitlab.com"]
+        bitbucket_cloud [label="bitbucket.com"]
```
Comment (unknwon) [inline doc/dev/architecture/architecture.dot]: It is `bitbucket.org`, though Atlassian has both and redirects `.com` to `.org`. 😛

**Comment by ryanslade** [comment]
How is the SVG generated? Is it something we can automate?

**Comment by tsenart** [comment]
> How is the SVG generated? Is it something we can automate?

It's generated with Graphviz. We can document the specific command for team mates who aren't familiar with it (i.e. `dot -Tsvg -o architecture.svg architecture.dot`)

**Comment by tsenart** [comment]
Final version (for now) with the fixes requested.

architecture

File: doc/dev/architecture/architecture.dot (hunk 4 above)
Comment (unknwon) [reply doc/dev/architecture/architecture.dot]: Missed this one?

**Comment by unknwon** [comment]
The color is legendary 😁

**Comment by unknwon** [comment]
This is ultra ultra approve (+7)!

**Pull request merged**